### DISCUSSION TITLE
Trying to use the composite command on Thousands of images on one common background keeping the original file name.

### DISCUSSION THREAD

**Discussion by soyebravat** [ImageMagick/ImageMagick#6069] [Help] [answered]
### ImageMagick version

7.1.0.6

### Operating system, version and so on

windows 11

### Description

I'm currently using this command:

magick background.jpeg null: ( *.png -resize 120x1200 ) -gravity center -geometry +0+0 -layers composite

Problem I have is it destroys my original file name and I am using it on over hundreds of images. Is there any way to batch apply without destroying the original file name? My main focus is in trying to keep the original  file name.

Much appreciated in advance.

**Comment by fmw42** [comment]
You have no output image specified in that command.  Please explain what you mean by "it destroys my original file name". The input should not be changed if you specify a proper output file.

**Comment by snibgo** [comment]
I think the OP wants the output filenames to be the same as the input filenames.

@soyebravat: Is your background 120x1200 pixels? If it is, then you can put the files in the opposite order and use a `filename:` construct to get the same filenames, like this:

```
magick ( *.png -resize 120x1200 ) null: background.jpeg -set filename:MYFILE "%%t" -gravity center -geometry +0+0 -compose DstOver -layers composite %[filename:MYFILE].png
```

Another option is to have a shell loop that calls `magick` once for every input file.

**Reply by soyebravat** [reply]
Hi, Ive been playing with the above command and its sorted the filename issue but threw me another issue. My command before which was :

magick background.jpeg null: ( *.png -resize 550x550 ) -gravity center -geometry +115-80 -layers composite result3.jpg
(this allowed a static  background image and all pngs to resize, center on the background image and adjust geometry but had the issue with not keeping the same filename)

magick ( *.png -resize 120x1200 ) null: background.jpeg -set filename:MYFILE "%%t" -gravity center -geometry +0+0 -compose DstOver -layers composite %[filename:MYFILE].png
(your command works prefectly on the filename, exactly what i want but the background and the foreground changes sizes and or the background is oversizing and cant seemd to under whats happening.)

Im hoping to have the filename as in your command but want the background static and adjust the size and gravity of the overlay pngs.

could you advise?

Much appreciated in advance.

**Comment by soyebravat** [comment]
Thanks ever so much for you replies,

I saved the below in a bat file:

magick ( *.png -resize 120x1200 ) null: background3.gif -set filename:MYFILE "%%t" -gravity center -geometry +0+0 -compose DstOver -layers composite %[filename:MYFILE].png



It executes perfectly but the filenames are saved as MYFILE]-1, MYFILE]-2, MYFILE]-3 

instead of the original bonsai, cat, car

am i doing something wrong?

thank you

**Comment by soyebravat** [comment]
I added an extra % works perfectly thank you sooo much!!!

highly appreciated :-)

**Comment by soyebravat** [comment]
Sorry have another question; 

I run the below line in cmd and works perfectly fine but in a .bat file its throwing errors any ideas why?

magick *.svg -fuzz 60% -fill VioletRed4  -opaque black -set filename:base "%[basename]" "%[filename:base]_Violet.png"

Thanks again

**Reply by fmw42** [reply]
In .bat files % has to be double (escaped) to %%

**Comment by snibgo** [comment]
> (your command works prefectly on the filename, exactly what i want but the background and the foreground changes sizes and or the background is oversizing and cant seemd to under whats happening.)

Sorry, I don't understand. Please show a background image and at least two foreground images.

**Comment by soyebravat** [comment]
So using this command: 


magick ( *.png -resize 120x1200 ) null: background6.jpeg -set filename:MYFILE "%%t" -gravity center -geometry +0+0 -compose DstOver -layers composite %%[filename:MYFILE].png

with this background:
![background6](https://user-images.githubusercontent.com/33333343/219905361-19fec727-5ddd-4e2f-9e62-dc67c113829b.jpeg)

this foreground:
![bruce lee](https://user-images.githubusercontent.com/33333343/219905449-131d8ae8-9dab-4df3-910f-07f376c5ffb6.png)

gives me this result:

![bruce lee_test file](https://user-images.githubusercontent.com/33333343/219905475-35fd1e12-3154-4e9c-b60a-bfc6714d3d5e.png)

instead of something like this which is what i'm after:

![example](https://user-images.githubusercontent.com/33333343/219905524-8e485c14-e8a4-44af-8e18-dd28ffc824fc.png)

It appears its cropping and zooming in to the background. I would like the changes only to be applied to the foreground image and not the background image as above. 

Much appreciated in advance!

**Comment by snibgo** [comment]
Your background is 10000x3330 pixels, and the foreground is a different size, even when resized. The resizing shrinks the forground image. Do you really want that? You can "`-extent`" the foreground to the same size as the background. Windows BAT syntax
```
magick ^
  ( *.png -resize 120x1200 -gravity Center -background None -extent 10000x3330 ) ^
  null: ^
  backgChair.jpeg ^
  -set filename:MYFILE "%%t" -gravity center -geometry +0+0 ^
  -compose DstOver -layers composite ^
  %%[filename:MYFILE].png
```
Beware: this will overwrite the foreground files.

If you have many *.png foreground files, you may run out of memory.

**Comment by soyebravat** [comment]
Sorry to be a pain but its almost working!

The entire script works almost perfectly as intended but the Geometry. Gravity works perfectly as it should, as does the size. When adding a reasonable value to the geometry, It doesn't budge at all. However, when putting a significantly large value to the geometry. It offsets with the background. Is there any way of fixing this? Apart from that, everything else appears to work perfectly.

So for the below image i added a significantly high value. I believe it was about 5000 both on the XY which gives me this:

![TEST2](https://user-images.githubusercontent.com/33333343/220186742-6b399e14-7937-4e4c-9aa8-1f114df29492.png)

When adding a reasonable value such as 100 on the x and y the result is below, It doesn't do anything and keeps it flush on the gravity set as northeast . This is playing with one or the other and both Geometries on the background and The PNG's.

on Some background files I'm trying to get a specific position which requires me to use geometry as well as gravity to get the desired result.

if this works it would be pretty awesome!!


![TEST](https://user-images.githubusercontent.com/33333343/220186971-d7cfc3dc-b102-4053-b35d-b19b16c778f4.png)

Much appreciated

**Reply by fmw42** [reply]
You do not say what your command line was. We need to see that to be able to comment.

**Reply by soyebravat** [reply]
This was the command, i've played with the geometry:

magick ^
  ( test.png -resize 750x750 -gravity northeast -geometry -50-100 -background None -extent 1600x1124 ) ^
  null: ^
  background2.png ^
  -set filename:MYFILE "%%t" -gravity northeast -geometry -50-100 ^
  -compose DstOver -layers composite ^
  %%[filename:MYFILE].png

**Reply by fmw42** [reply]
I am unaware that -extent -geometry sensitive.  But I think you can add offsets to the extent argument. Try

magick ^
( test.png -resize 750x750 -gravity northeast -background None -extent 1600x1124-50-100 ) ^
null: ^
background2.png ^
-set filename:MYFILE "%%t" -gravity northeast -geometry -50-100 ^
-compose DstOver -layers composite ^
%%[filename:MYFILE].png

However,, I am not sure about the negative values. But give it a try. Perhaps you meant +50+100?  Untested!

**Comment by soyebravat** [comment, marked as answer]
Absolutley smashing!!!! that works perfectly and does everything i want it to do. Thanks to you!!! thank you Snibgo and fmw42!!!
